crawler/urlJobProcessor.js

```diff
@@ -1,36 +1,106 @@
 
 const probes = [];
 
-function configure(/* opts */) {
-  // TODO: Set up probes. They'll be specified somehow in opts and we
-  // should be able to load each probe individually. Maybe something
-  // like `require('./probe-' + probeName + '.js');`
+function configure(opts) {
+  if (!opts.probes || !opts.probes.length) {
+    return Promise.reject(new Error('At least one urlJobProcessor probe must be specified'));
+  }
+
+  const ret = [Promise.resolve()];
+
+  opts.probes.forEach((probeOpts) => {
+    if (!probeOpts.name) {
+      return Promise.reject(new Error('Probe without name encountered'));
+    }
+
+    // We allow the probe to be specified by name or by path. If specified by name,
+    // we expect it to be a known subdir.
+    const path = probeOpts.path || `./urlJobProcessorProbes/${probeOpts.name}.js`;
+    const probeModule = require(path).default;
+
+    if (!probeModule.runProbe) {
+      return Promise.reject(new Error(`Probe ${probeOpts.name} does not implement runProbe`));
+    }
+
+    let p = Promise.resolve();
+    if (probeModule.configure) {
+      p = probeModule.configure(probeOpts.opts || {});
+    }
+
+    ret.push(p.then(() => {
+      probes.push({
+        name: probeOpts.name,
+        probeFn: probeModule.runProbe,
+      });
+    }));
+  });
+
+  return Promise.all(ret);
 }
 
-function fetchAllResources(url) {
-  const ret = { html: '', js: '', sw: '' };
-  return fetch(`http://${url}`).then((response) => {
-    return response.text().then((htmlText) => {
-      ret.html += htmlText;
+function fetchAllResources(urlArg) {
+  const url = urlModule.parse(`http://${urlArg}`);
+
+  // TODO: Pass user agent in headers
+  return fetch(`${url.href}`, { headers: '' }).then((mainResponse) => {
+    // We pass along the response object that was returned after
+    // our fetch of the main page. As an optimization, we don't
+    // clone the response here. Instead, we pass the body text
+    // in a separate arg.
+    const responses = [mainResponse];
+    return mainResponse.text().then((htmlText) => {
+      const $ = cheerio.load(htmlText);
+      const scripts = [];
+      const promises = [];
+
+      $('script').each((index, elem) => {
+        if (!$(elem).attr('src')) {
+          // For inline scripts, just append them to the array of scripts
+          // that we'll pass to the probes
+          scripts.push($(elem).text());
+        } else {
+          // For scripts with a `src` attribute, we must fetch them here
+          let src = $(elem).attr('src');
+          if (src[0] === '/') {
+            // Deal with relative paths in `src` attributes
+            src = `${url.protocol}//${url.host}${src}`;
+          }
+          promises.push(fetch(src).then((scriptResponse) => {
+            return scriptResponse.text();
+          }).then((jsText) => {
+            return scripts.push(jsText);
+          }));
+        }
+      });
+
+      return Promise.all(promises).then(() => {
```

Comment:
> Is the intention that missing JS file will fail the whole crawling for this site, the site will go into the re-test queue a few times and then get removed. Is that the right punishment for a missing JS file?

It's a trade-off. When we're unable to fetch a particular JS file, we either put data in our data store that is inaccurate or we refuse to add an entry for that site. I'm leaning toward refusing to add an entry rather than knowingly pollute our data store. We might end up with some gaps for some sites but I think that's better than having bad data that we could have avoided.